README.md

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/vineclust)](https://CRAN.R-project.org/package=vineclust)
[![R-CMD-check](https://github.com/oezgesahin/vineclust/workflows/R-CMD-check/badge.svg)](https://github.com/oezgesahin/vineclust/actions)
[![codecov](https://codecov.io/gh/oezgesahin/vineclust/branch/main/graph/badge.svg?token=EUTCCS3B5U)](https://codecov.io/gh/oezgesahin/vineclust)
<!-- badges: end -->

# vineclust: Model-Based Clustering with Vine Copulas

An R package that fits vine copula based mixture model distributions to
the continuous data for a given number of components as proposed in
[VCMM algorithm](https://arxiv.org/pdf/2102.03257.pdf) and use its
results for clustering.

It depends on [VineCopula](https://github.com/tnagler/VineCopula),
[fGarch](https://github.com/cran/fGarch),
[mclust](https://github.com/cran/mclust), and
[univariateML](https://github.com/JonasMoss/univariateML).

## Installation

You can install the development version from
[GitHub](https://github.com/oezgesahin) with:

``` r
# install.packages("remotes")
remotes::install_github("oezgesahin/vineclust")
```

## Package overview

Below is an overview of some functions and features.

  - `vcmm()`: fits vine copula based mixture model distributions to the
    continuous data for a given number of components. Returns an object
    of class `vcmm_res()`. The class has the following methods:
      - `print`: a brief overview of the model statistics.
      - `summary`: list of fitted model components, including selected
        vine tree structures, bivariate copula families, univariate
        marginal distributions, and estimated parameters.
  - `dvcmm(), rvcmm()`: density and random generation for the vine
    copula based mixture model distributions.

### Bivariate copula families

This package works with a wide range of parametric bivariate copula
families for bivariate or multivariate clustering using vine copulas.
Specifically, it allows fitting elliptical (Gaussian, Student-t) and
Archimedean (Clayton, Gumbel, Frank, Joe, BB1, BB6, and BB8) copulas
with their possible 90, 180, 270 degrees rotations to cover a large
range of dependence patterns. Their encoding is detailed on
[VineCopula](https://github.com/tnagler/VineCopula).

### Univariate marginal distributions

This package currently includes following unimodal univariate marginal
distributions.

  - `cauchy(a,b)`: Cauchy distribution with location parameter a and
    scale parameter b,
  - `gamma(a,b)`: gamma distribution with shape parameter a and rate
    parameter b,
  - `llogis(a,b)`: log-logistic distribution with shape parameter a and
    rate parameter b,
  - `lnorm(a,b)`: log-normal distribution with mean parameter a and
    standard deviation parameter b on the logarithmic scale,
  - `logis(a,b)`: logistic distribution with location parameter a and
    scale parameter b,
  - `norm(a,b)`: normal distribution with mean parameter a and standard
    deviation parameter b,
  - `snorm(a,b,c)`: skew normal distribution with location parameter a,
    scale parameter b, and skewness parameter c.
  - `std(a,b,c)`: Student’s t distribution with location parameter a,
    scale parameter b, and shape parameter c,
  - `sstd(a,b,c,d)`: skew Student’s t distribution with location
    parameter a, scale parameter b, shape parameter c, and skewness
    parameter d.

### Initial partition methods

This package currently implements following partition approaches to have
starting values.

  - `kmeans`: performs k-means clustering (Hartigan-Wong) on given data
    after scaling,
  - `hcVVV`: performs model-based hierarchical clustering on given data
    after scaling,
  - `gmm`: performs model-based clustering with Gaussian mixture models
    on given data.

## Usage

``` r
library(vineclust)
# data from UCI Machine Learning Repository 
data_wisc <- read.csv("http://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/wdbc.data", header = FALSE)
```

### Model fitting

``` r
 # R-vine copula based mixture model with total components 2
fit <- vcmm(data=data_wisc[,c(15,27,29,30)], total_comp=2)
 # Model statistics
print(fit)
#> logLik = 2159.11   BIC = -4108.88   ICL = -4068.3   npars = 33   initial clustering = kmeans   ECM iterations = 4
# Fitted vine copula distributions
summary(fit) 
#> $margins
#>      [,1]          [,2]         
#> [1,] "Lognormal"   "Lognormal"  
#> [2,] "Lognormal"   "Gamma"      
#> [3,] "Lognormal"   "Skew Normal"
#> [4,] "Skew Normal" "Normal"     
#> 
#> $marginal_pars
#> , , 1
#> 
#>           [,1]       [,2]       [,3]       [,4]
#> [1,] 1.3257284 -1.9132582 -0.8083077 0.18916084
#> [2,] 0.5085319  0.1325288  0.3702962 0.03873749
#> [3,]        NA         NA         NA 1.81086926
#> [4,]        NA         NA         NA         NA
#> 
#> , , 2
#> 
#>           [,1]      [,2]       [,3]       [,4]
#> [1,] 0.6513504  42.84792  0.1607200 0.07370799
#> [2,] 0.3849773 347.42118  0.1222864 0.03358325
#> [3,]        NA        NA 13.0380864         NA
#> [4,]        NA        NA         NA         NA
#> 
#> 
#> $copula
#> , , 1
#> 
#>      [,1] [,2] [,3] [,4]
#> [1,]    0    0    0    0
#> [2,]    1    0    0    0
#> [3,]   24    1    0    0
#> [4,]    5    1   20    0
#> 
#> , , 2
#> 
#>      [,1] [,2] [,3] [,4]
#> [1,]    0    0    0    0
#> [2,]   33    0    0    0
#> [3,]   33   23    0    0
#> [4,]   10    3   14    0
#> 
#> 
#> $copula_first_par
#> , , 1
#> 
#>            [,1]       [,2]  [,3] [,4]
#> [1,]  0.0000000  0.0000000 0.000    0
#> [2,] -0.3781992  0.0000000 0.000    0
#> [3,] -1.0644572 -0.2277140 0.000    0
#> [4,]  2.1769280  0.3734696 5.682    0
#> 
#> , , 2
#> 
#>             [,1]        [,2]    [,3] [,4]
#> [1,]  0.00000000  0.00000000 0.00000    0
#> [2,] -0.09652639  0.00000000 0.00000    0
#> [3,] -0.19082564 -0.07882284 0.00000    0
#> [4,]  1.30499379  0.41096477 2.66261    0
#> 
#> 
#> $copula_second_par
#> , , 1
#> 
#>      [,1] [,2]      [,3] [,4]
#> [1,]    0    0 0.0000000    0
#> [2,]    0    0 0.0000000    0
#> [3,]    0    0 0.0000000    0
#> [4,]    0    0 0.6308378    0
#> 
#> , , 2
#> 
#>           [,1] [,2] [,3] [,4]
#> [1,] 0.0000000    0    0    0
#> [2,] 0.0000000    0    0    0
#> [3,] 0.0000000    0    0    0
#> [4,] 0.9357103    0    0    0
#> 
#> 
#> $vine_structure
#> , , 1
#> 
#>      [,1] [,2] [,3] [,4]
#> [1,]    2    0    0    0
#> [2,]    1    1    0    0
#> [3,]    4    3    3    0
#> [4,]    3    4    4    4
#> 
#> , , 2
#> 
#>      [,1] [,2] [,3] [,4]
#> [1,]    1    0    0    0
#> [2,]    3    2    0    0
#> [3,]    2    3    3    0
#> [4,]    4    4    4    4
#> 
#> 
#> $mixture_probs
#> [1] 0.3524161 0.6475839
# Evaluate the density of the fitted model at (2.747, 0.1467, 0.13, 0.05334)
RVMs_fitted <- list()
RVMs_fitted[[1]] <- VineCopula::RVineMatrix(Matrix=fit$output$vine_structure[,,1],
                        family=fit$output$bicop_familyset[,,1],
                        par=fit$output$bicop_param[,,1],
                        par2=fit$output$bicop_param2[,,1])
RVMs_fitted[[2]] <- VineCopula::RVineMatrix(Matrix=fit$output$vine_structure[,,2],
                        family=fit$output$bicop_familyset[,,2],
                        par=fit$output$bicop_param[,,2],
                        par2=fit$output$bicop_param2[,,2])
dvcmm(c(2.747, 0.1467, 0.13, 0.05334), fit$output$margin, fit$output$marginal_param, RVMs_fitted, fit$output$mixture_prob)
#> [1] 26.70683
```

``` r
# C-vine copula based mixture model
fit_cvine <- vcmm(data=data_wisc[,c(15,27,29,30)], total_comp=2, is_cvine=1)  
# Confusion matrix w.r.t. true classification
table(fit_cvine$cluster, data_wisc$V2) 
#>    
#>       B   M
#>   1  32 192
#>   2 325  20
```

``` r
# Fit only bivariate Clayton copula for pairs of variables in both components
fit_clayton <- vcmm(data=data_wisc[,c(15,27,29,30)], total_comp=2, bicop=c(3))  
```

``` r
# Fix vine tree structures of both components 
fit_fix_vinestr <- vcmm(data=data_wisc[,c(15,27,29,30)], total_comp=2,
                        vinestr=matrix(c(1,2,3,4,0,2,4,3,0,0,4,3,0,0,0,3),4,4)) 
```

``` r
# Run ECM iterations shorter with a smaller threshold than the default threshold
fit_sthr <- vcmm(data=data_wisc[,c(15,27,29,30)], total_comp=2, threshold=0.001)  
```

``` r
# Use a different initial partition approach from k-means
fit_best_init <- vcmm(data=data_wisc[,c(15,27,29,30)], total_comp=2, methods=c("gmm"))  
```

### Simulation

``` r
# Simulation setup given in Section 5.2 of the paper at https://arxiv.org/pdf/2102.03257.pdf
dims <- 3
obs <- c(500,500) 
RVMs <- list()
RVMs[[1]] <- VineCopula::RVineMatrix(Matrix=matrix(c(1,3,2,0,3,2,0,0,2),dims,dims),
                        family=matrix(c(0,3,4,0,0,14,0,0,0),dims,dims),
                        par=matrix(c(0,0.8571429,2.5,0,0,5,0,0,0),dims,dims),
                        par2=matrix(sample(0, dims*dims, replace=TRUE),dims,dims)) 
RVMs[[2]] <- VineCopula::RVineMatrix(Matrix=matrix(c(1,3,2,0,3,2,0,0,2), dims,dims),
                        family=matrix(c(0,6,5,0,0,13,0,0,0), dims,dims),
                        par=matrix(c(0,1.443813,11.43621,0,0,2,0,0,0),dims,dims),
                        par2=matrix(sample(0, dims*dims, replace=TRUE),dims,dims))
margin <- matrix(c('Normal', 'Gamma', 'Lognormal', 'Lognormal', 'Normal', 'Gamma'), 3, 2) 
margin_pars <- array(0, dim=c(2, 3, 2))
margin_pars[,1,1] <- c(1, 2)
margin_pars[,1,2] <- c(1.5, 0.4)
margin_pars[,2,1] <- c(1, 0.2)
margin_pars[,2,2] <- c(18, 5)
margin_pars[,3,1] <- c(0.8, 0.8)
margin_pars[,3,2] <- c(1, 0.2)
x_data <- rvcmm(dims, obs, margin, margin_pars, RVMs)
```

## Contact

Please contact <ozge.sahin@tum.de> if you have any questions.

## References

Sahin, {"O}., & Czado, C. (2021). Vine copula mixture models and
clustering for non-gaussian data. Econometrics and Statistics.
<doi:10.1016/j.ecosta.2021.08.011>.
[preprint](https://arxiv.org/pdf/2102.03257.pdf),
[article](https://doi.org/10.1016/j.ecosta.2021.08.011)


NEWS.md
# vineclust 0.1.0

* First release.


tests/testthat.R
library(testthat)
library(vineclust)

test_check("vineclust")


tests/testthat/test-vcmm.R
# simulate data
set.seed(111)
cl1 <- sapply(1:2, function(i) rnorm(100, -10, 5))
cl2 <- sapply(1:2, function(i) rnorm(100, 10, 5))
x_data <- data.frame(rbind(cl1, cl2))

test_that("catches incorrect input", {
  expect_error(vcmm(x_data[,1], total_comp=2))
  expect_error(vcmm(matrix(2,2,2,'c'),total_comp=2))
  expect_error(vcmm(matrix(2,2,2,'c'),total_comp=2))
  expect_error(vcmm(matrix(2,2,2,NA),total_comp=2))
  expect_error(vcmm(total_comp=2))
  expect_error(vcmm(x_data))
  expect_error(vcmm(x_data, total_comp=2, is_cvine=2))
  expect_error(vcmm(x_data, total_comp=2, trclevel='xyz'))
  expect_error(vcmm(x_data, total_comp=2, trclevel=9999))
  expect_error(vcmm(x_data, total_comp=2, mar='whatisthis'))
  expect_error(vcmm(x_data, total_comp=2, mar=1))
  expect_error(vcmm(x_data, total_comp=2, bicop='xyz'))
  expect_error(vcmm(x_data, total_comp=2, bicop=9999))
  expect_error(vcmm(x_data, total_comp=2, methods = "xyz"))
  expect_error(vcmm(x_data, total_comp=2, methods = 9999))
  expect_error(vcmm(x_data, total_comp=2, threshold=-1))
  expect_error(vcmm(x_data, total_comp=2, threshold='xyz'))
  expect_error(vcmm(x_data, total_comp=2, maxit=-1))
  expect_error(vcmm(x_data, total_comp=2, maxit='xyz'))
})

fit <- vcmm(x_data, 2)

test_that("returns correct 'vcmm_res' object", {
  expect_s3_class(fit, "vcmm_res")
  expect_identical(
    names(fit),
    c(
      "output", "cluster"
    )
  )
})

test_that("proper print/summary generic functions", {
  expect_output(print(fit))
  svcmm <- summary(fit)
  expect_type(svcmm$margins, "character")
  expect_type(svcmm$marginal_pars, "double")
  expect_type(svcmm$copula, "double")
  expect_type(svcmm$copula_first_par, "double")
  expect_type(svcmm$copula_second_par, "double")
  expect_type(svcmm$vine_structure, "double")
  expect_type(svcmm$mixture_probs, "double")
})

test_that("fixed mar/bicop selection works", {
  expect_type(
    vcmm(x_data, total_comp=2, mar='norm', bicop=1)$output$margin,
    "character"
  )
})

test_that("other threshold works", {
  expect_type(
    vcmm(x_data, total_comp=2, threshold=0.1)$output$margin,
    "character"
  )
})


test_that("other maxit works", {
  expect_type(
    vcmm(x_data, total_comp=2, maxit=2)$output$margin,
    "character"
  )
})



tests/testthat/test-margins.R
# simulate data
set.seed(111)
x_data <- rgamma(100, 1, 2)
fit <- fit_margin(x_data, 1, NA)
pars <- c(1,2)

test_that("marginal fitting returns correct output", {
  expect_identical(
    names(fit),
    c(
      "par_mar", "fam"
    )
  )
})

test_that("marginal fitting work", {
  expect_type(fit$par_mar, "double")
  expect_type(fit$fam, "character")
})


test_that("density functions for margins work", {
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Cauchy", pars, "pdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Gamma", pars, "pdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Logistic", pars, "pdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Loglogistic", pars, "pdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Lognormal", pars, "pdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Normal", pars, "pdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Skew Normal", c(1,2,1), "pdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Student-t", c(1,2,3), "pdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Skew Student-t", c(1,2,3,4), "pdf")),
    0
  )
})

test_that("distribution functions for margins work", {
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Cauchy", pars, "cdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Gamma", pars, "cdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Logistic", pars, "cdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Loglogistic", pars, "cdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Lognormal", pars, "cdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Normal", pars, "cdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Skew Normal", c(1,2,1), "cdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Student-t", c(1,2,3), "cdf")),
    0
  )
  expect_gte(
    min(pdf_cdf_quant_margin(x_data, "Skew Student-t", c(1,2,3,4), "cdf")),
    0
  )
  expect_lte(
    max(pdf_cdf_quant_margin(x_data, "Cauchy", pars, "cdf")),
    1
  )
  expect_lte(
    max(pdf_cdf_quant_margin(x_data, "Gamma", pars, "cdf")),
    1
  )
  expect_lte(
    max(pdf_cdf_quant_margin(x_data, "Logistic", pars, "cdf")),
    1
  )
  expect_lte(
    max(pdf_cdf_quant_margin(x_data, "Loglogistic", pars, "cdf")),
    1
  )
  expect_lte(
    max(pdf_cdf_quant_margin(x_data, "Lognormal", pars, "cdf")),
    1
  )
  expect_lte(
    max(pdf_cdf_quant_margin(x_data, "Normal", pars, "cdf")),
    1
  )
  expect_lte(
    max(pdf_cdf_quant_margin(x_data, "Skew Normal", c(1,2,1), "cdf")),
    1
  )
  expect_lte(
    max(pdf_cdf_quant_margin(x_data, "Student-t", c(1,2,3), "cdf")),
    1
  )
  expect_lte(
    max(pdf_cdf_quant_margin(x_data, "Skew Student-t", c(1,2,3,4), "cdf")),
    1
  )
})

test_that("quantile functions for margins work", {
  expect_type(
    pdf_cdf_quant_margin(x_data[1], "Cauchy", pars, "quant"),
    "double"
  )
  expect_type(
    pdf_cdf_quant_margin(x_data[1], "Gamma", pars, "quant"),
    "double"
  )
  expect_type(
    pdf_cdf_quant_margin(x_data[1], "Logistic", pars, "quant"),
    "double"
  )
  expect_type(
    pdf_cdf_quant_margin(x_data[1], "Loglogistic", pars, "quant"),
    "double"
  )
  expect_type(
    pdf_cdf_quant_margin(x_data[1], "Lognormal", pars, "quant"),
    "double"
  )
  expect_type(
    pdf_cdf_quant_margin(x_data[1], "Normal", pars, "quant"),
    "double"
  )
  expect_type(
    pdf_cdf_quant_margin(x_data[1], "Skew Normal", c(1,2,1), "quant"),
    "double"
  )
  expect_type(
    pdf_cdf_quant_margin(x_data[1], "Student-t", c(1,2,3), "quant"),
    "double"
  )
  expect_type(
    pdf_cdf_quant_margin(x_data[1], "Skew Student-t", c(1,2,3,1), "quant"),
    "double"
  )
})


R/vcmm.R
#' Model-Based Clustering with Vine Copulas
#'
#' It fits vine copula based mixture model distributions to the continuous data
#' for a given number of components as described in Sahin and Czado (2021)
#' and use its results for clustering.
#'
#' @param data 	A matrix or data frame of observations. Categorical/discrete variables not (yet) allowed.
#' If a matrix or data frame, rows correspond to observations (i) and columns correspond to variables (p).
#' @param total_comp An integer specifying the numbers of mixture components (clusters)
#' @param is_cvine An integer specifying if the type of components' vine tree structure is C-vine
#' before/after the ECM phase of clustering.
#' * 0 = R-vine (default)
#' * 1 = C-vine
#' @param vinestr A matrix specifying vine tree structures before/after the ECM phase of clustering.
#' The default is automatic selection.
#' \link[VineCopula]{RVineMatrixCheck} checks for a valid R-vine matrix.
#' @param trunclevel An integer showing the level of truncation for vine tree structures before the ECM phase of clustering.
#' The default is 1.
#' @param mar A vector of character strings indicating the parametric univariate marginal distributions
#' to be fitted before/after the ECM phase of clustering.
#' The default is c('cauchy','gamma','llogis','lnorm','logis','norm','snorm','std', 'sstd').
#' Other distributions not (yet) allowed.
#' @param bicop A vector of integers denoting the parametric bivariate copula families to be fitted
#' before/after the ECM phase of clustering.
#' The default is c(1,2,3,4,5,6,7,8,10,13,14,16,17,18,20,23,24,26,27,28,30,33,34,36,37,38,40).
#' \link[VineCopula]{BiCop} describes the available families with their specifications.
#' @param methods A vector of character strings indicating initial clustering method(s) to have a partition
#' for model selection before the ECM phase of clustering. Current options:
#' * 'kmeans' (default)
#' * c('kmeans', 'gmm', 'hcVVV')
#' @param threshold A numeric, stopping the ECM phase of clustering. The default is 1e-4.
#' @param maxit An integer, specifying the maximum number of iterations in the CM-step 2 optimization. The default is 10.
#'
#' @return An object of class vcmm result. It contains the elements
#' \describe{
#' \item{cluster}{the vector with the classification of observations}
#' \item{output}{a list containing the fitted VCMM.}
#' } Use `print.vcmm_res()` to obtain log-likelihood, BIC, ICL, number of estimated parameters, initial clustering method used
#'  and total number of ECM iterations for the fitted VCMM. `summary.vcmm_res()` shows the fitted vine tree structures and
#'  univariate marginal distributions, bivariate copula families with the estimated parameters, as well as
#'  mixture proportions of each component.
#'
#' @references
#' Sahin and Czado (2021), Vine copula mixture models and clustering for non-Gaussian data, Econometrics and Statistics.
#' doi: 10.1016/j.ecosta.2021.08.011
#'
#' @seealso [dvcmm()], [rvcmm()]
#'
#' @examples
#' # Simulate 3-dimensianal data from parametric vine copula based mixture model
#' # with 2 components on x-scale. Each component has 500 observations.
#' dims <- 3
#' obs <- c(500,500)
#' RVMs <- list()
#' RVMs[[1]] <- VineCopula::RVineMatrix(Matrix=matrix(c(1,3,2,0,3,2,0,0,2),dims,dims),
#'                         family=matrix(c(0,3,4,0,0,14,0,0,0),dims,dims),
#'                         par=matrix(c(0,0.8571429,2.5,0,0,5,0,0,0),dims,dims),
#'                         par2=matrix(sample(0, dims*dims, replace=TRUE),dims,dims))
#' RVMs[[2]] <- VineCopula::RVineMatrix(Matrix=matrix(c(1,3,2,0,3,2,0,0,2), dims,dims),
#'                         family=matrix(c(0,6,5,0,0,13,0,0,0), dims,dims),
#'                         par=matrix(c(0,1.443813,11.43621,0,0,2,0,0,0),dims,dims),
#'                         par2=matrix(sample(0, dims*dims, replace=TRUE),dims,dims))
#' margin <- matrix(c('Normal', 'Gamma', 'Lognormal', 'Lognormal', 'Normal', 'Gamma'), 3, 2)
#' margin_pars <- array(0, dim=c(2, 3, 2))
#' margin_pars[,1,1] <- c(1, 2)
#' margin_pars[,1,2] <- c(1.5, 0.4)
#' margin_pars[,2,1] <- c(1, 0.2)
#' margin_pars[,2,2] <- c(18, 5)
#' margin_pars[,3,1] <- c(0.8, 0.8)
#' margin_pars[,3,2] <- c(1, 0.2)
#' x_data <- rvcmm(dims, obs, margin, margin_pars, RVMs)
#'
#' # Example-1: fit parametric 3 dimensional R-vine copula based mixture model with 2 components,
#' # using all univariate marginal distributions and bivariate copula families allowed,
#' # initial partition of k-means
#' \dontrun{
#' fit <- vcmm(x_data[,1:3], total_comp = 2)
#' print(fit)
#' summary(fit)
#' table(x_data$comp_id, fit$cluster)
#' # evaluate the density at (X1, X2, X3) = (1,2,3) for the fitted vcmm
#' # after encoding fitted vine copula
#' RVMs_fitted <- list()
#' RVMs_fitted[[1]] <- VineCopula::RVineMatrix(Matrix=fit$output$vine_structure[,,1],
#'                         family=fit$output$bicop_familyset[,,1],
#'                         par=fit$output$bicop_param[,,1],
#'                         par2=fit$output$bicop_param2[,,1])
#' RVMs_fitted[[2]] <- VineCopula::RVineMatrix(Matrix=fit$output$vine_structure[,,2],
#'                         family=fit$output$bicop_familyset[,,2],
#'                         par=fit$output$bicop_param[,,2],
#'                         par2=fit$output$bicop_param2[,,2])
#' dvcmm(c(1, 2, 3), fit$output$margin, fit$output$marginal_param,
#' RVMs_fitted, fit$output$mixture_prob)
#' }
#'
#' @export
#'
#' @import VineCopula
#' @import mclust
#' @import univariateML
#' @importFrom parallel mclapply
#' @importFrom fGarch psnorm dsnorm pstd dstd psstd dsstd
#' @importFrom stats dgamma dlnorm dlogis dnorm dcauchy kmeans optim pgamma plnorm plogis pnorm pcauchy sd

vcmm <- function(data, total_comp, is_cvine=NA, vinestr=NA, trunclevel=1, mar=NA, bicop=NA,
                 methods=c('kmeans'),  threshold=0.0001, maxit=10){
  initial_df_check(data)
  initial_args_check(data, total_comp, is_cvine, vinestr, trunclevel, mar, bicop,
                     methods, threshold, maxit)
  final_cvine <- is_cvine
  final_vinestr <- vinestr
  final_trunclevel <- NA
  final_mar <- mar
  final_bicop <- bicop
  winner_bic <- 1000000
  for(method in methods){
    initial_out <- initial_clustering(data, total_comp, is_cvine, vinestr, trunclevel, mar, bicop, method)
    marginal_params <- initial_out$marginal_params
    marginal_fams <- initial_out$marginal_fams
    u_data <- initial_out$u_data
    vine_structures <- initial_out$vine_structures
    family_sets <- initial_out$family_sets
    cop_params <- initial_out$cop_params
    cop_params_2 <- initial_out$cop_params_2
    mix_probs <- initial_out$mix_probs
    total_obs <- dim(data)[1]
    total_features <- dim(data)[2]
    iteration <- 1
    loglik_res <- vector()
    cond <- TRUE
    while(cond==TRUE){
      rvine_densities <- matrix(0, total_obs, total_comp)
      total_margin_dens <- matrix(0, total_obs, total_comp)
      margin_densities <- array(0, dim=c(total_obs, total_features, total_comp))
      lik_points <- matrix(0,dim(data)[1],total_comp)
      rvine_densities <-sapply(1:total_comp, function(j) rvine_density(u_data[,,j], vine_structures[,,j], family_sets[,,j],
                                                                          cop_params[,,j], cop_params_2[,,j]))
      for(j in 1:total_comp){
        density <- rep(1, dim(margin_densities)[1])
        margin_densities[,,j]<-sapply(1:total_features, function(x) pdf_cdf_quant_margin(data[,x],marginal_fams[x,j],
                                                                                         marginal_params[,x,j], 'pdf'))
        for(t in 1:total_features){
          density <- density * margin_densities[,t,j]
        }
        total_margin_dens[,j] <- density
      }
      lik_points <-sapply(1:total_comp, function(j) mix_probs[j]*total_margin_dens[,j]*rvine_densities[,j])
      lik_per_obs <- apply(lik_points, 1, sum)
      lik_per_obs[which(lik_per_obs == 0)] <- 1e-100
      loglik <- sum(log(lik_per_obs))
      loglik_res[iteration] <- loglik
      if(iteration > 2){
        if ((abs(loglik_res[iteration]-loglik_res[iteration-1])/abs(loglik_res[iteration-1]))
            <= threshold){
          cond <- FALSE
          break
        }
      }
      #if(((iteration-1) - floor((iteration-1)/10)*10)==0 & (iteration-1) > 0) cat(iteration-1, "ECM iterations are complete", "\n")
      #E-step
      z_values <- lik_points/rep(lik_per_obs, total_comp)
      #CM-steps:
      #CM-step 1
      mix_probs <- CM_step_mixture_probs(z_values)
      #CM-step 2 and 3
      CMS <- parallel::mclapply(1:total_comp, function(x) CM_steps(data, vine_structures[,,x], family_sets[,,x], cop_params[,,x],
                                                         cop_params_2[,,x], z_values[,x], marginal_fams[,x], marginal_params[,,x],
                                                         maxit))
      for(j in 1:total_comp){
        marginal_params[,,j] <- CMS[[j]]$marginal_par
        cop_params[,,j] <-  CMS[[j]]$cop_param
        cop_params_2[,,j] <- CMS[[j]]$cop_param_2
        u_data[,,j] <- CMS[[j]]$u_data
      }
      iteration <- iteration + 1
    }
    iteration <- iteration - 1
    mix_probs <- CM_step_mixture_probs(z_values)
    final_out <- final_selection(data, total_comp, final_cvine, final_vinestr, final_trunclevel, mix_probs, z_values,
                                 iteration, method, final_mar, final_bicop)
    vcmm_bic <- final_out$bic
    if(vcmm_bic < winner_bic){
      winner_bic <- vcmm_bic
      out <- final_out
      vcmm_class <- apply(out$z_values,1,function(x) which(x==max(x)))
    }
  }
  winner_bic <- round(winner_bic, 0)
  out_list <- list("output"=out, "cluster"=vcmm_class)
  class(out_list) <- "vcmm_res"
  out_list
}

#' @export
print.vcmm_res <- function(x, ...) {
  fit_info(x)
  invisible(x)
}

#' @export
summary.vcmm_res <- function(object, ...) {
  list(
    margins = object$output$margin,
    marginal_pars =  object$output$marginal_param,
    copula = object$output$bicop_familyset,
    copula_first_par = object$output$bicop_param,
    copula_second_par = object$output$bicop_param2,
    vine_structure =  object$output$vine_structure,
    mixture_probs = object$output$mixture_prob
  )
}



